### PR TITLE
Feat/plugin install

### PR DESCRIPTION
**Pull request by onerandomusername**
## Relevant Issues
Works on #63 
## Description
Adds a system to install plugins, and read them from a respository. A sample respository can be found here: [onerandomusername/modmal-addons](https://github.com/onerandomusername/modmail-addons)

### CODE REVIEW

**Comment by onerandomusername** [comment]
This seems to be ready for review now.

**Comment by onerandomusername** [comment]
This pr is now feature frozen.

The bot not using the cached downloaded files will be resolved in a seperate pr.

**Comment by onerandomusername** [comment]
This still depends on solving the docker problem, however the rest of the code can be review.

**Review by Shivansh-007 — CHANGES_REQUESTED** [review]
Haven't tested or read the docs yet

File: modmail/addons/models.py [hunk 1]
```diff
@@ -0,0 +1,264 @@
...
+        source.user = user
+        source.reflike = reflike
+        source.githost = githost
+        source.githost_api = Host
+        return source
+
+    @classmethod
+    def from_zip(cls, url: str) -> AddonSource:
+        """Create an AddonSource from a zip file."""
+        match = re.match(r"^(?P<url>(?:https?:\/\/)?(?P<domain>.*\..+?)\/(?P<path>.*\.zip))$", url)
```
[the remark this answers is not in the document]
Comment (onerandomusername) [reply modmail/addons/models.py]: Removed the regex in 6e349943cc8bf

**Comment by onerandomusername** [comment]
Linting workflows are going to fail until this is merged with main again due to how github merges then and a dependency mismatch. Given that there's so many merge commits already, I'm holding off on a merge until later.

File: modmail/__init__.py [hunk 2]
```diff
@@ -1,12 +1,18 @@
...
 from pathlib import Path
 
 import coloredlogs
 
 from modmail.log import ModmailLogger
 
 
+# On Windows, the selector event loop is required for aiodns.
+if os.name == "nt":
+    asyncio.set_event_loop_policy(asyncio.WindowsSelectorEventLoopPolicy())
```
[the remark this answers is not in the document]
Comment (onerandomusername) [reply modmail/__init__.py:14]: Moved this to #125 and implemented the change there.

File: modmail/__init__.py [hunk 3]
```diff
@@ -58,3 +64,9 @@
 logging.getLogger("websockets").setLevel(logging.ERROR)
 # Set asyncio logging back to the default of INFO even if asyncio's debug mode is enabled.
 logging.getLogger("asyncio").setLevel(logging.INFO)
+
+# now that the logger is configured, we can import the bot for using as all and typing
+from modmail.bot import ModmailBot  # noqa: E402
+
+
+__all__ = [ModmailBot, ModmailLogger]
```
[the remark this answers is not in the document]
Comment (onerandomusername) [reply modmail/__init__.py:72]: Actually, removing this.

File: modmail/addons/converters.py [hunk 4]
```diff
@@ -0,0 +1,67 @@
+from __future__ import annotations
```
[the remark this answers is not in the document]
Comment (onerandomusername) [reply modmail/addons/converters.py:1]: If its removed, it just means that the TYPE_CHECKING if either needs to be removed, or the imports in said if need to be stringifyed where they are used.

Although....

File: modmail/addons/models.py [hunk 5]
```diff
@@ -0,0 +1,264 @@
...
+        all_fully_matched_plugins: Set[Plugin] = set()
+        all_partially_matched_plugins: Dict[Plugin, float] = dict()
+        for res in result:
+            all_partially_matched_plugins[arg_mapping[res[0]]] = res[1]
+
+            if res[1] == 100:
+                all_fully_matched_plugins.add(arg_mapping[res[0]])
+
+        if len(all_fully_matched_plugins) != 1:
+            suggested = ""
```
[the remark this answers is not in the document]
Comment (onerandomusername) [reply modmail/addons/models.py:253]: all partially matched plugins is a superset of all fully matched, see the code above.

File: modmail/addons/plugins.py [hunk 6]
```diff
@@ -0,0 +1,370 @@
...
+
+
+async def install_dependencies(plugin: Plugin) -> str:
+    """Installs provided dependencies from a plugin."""
+    # check if there are any plugins to install
+    if not len(plugin.dependencies):
+        return
+
+    if PYTHON_INTERPRETER is None:
+        raise FileNotFoundError("Could not locate python interpreter.")
```
[the remark this answers is not in the document]
Comment (onerandomusername) [reply modmail/addons/plugins.py:76]: This will be caught by the error handler and not shown to the end user.

File: modmail/addons/plugins.py [hunk 7]
```diff
@@ -0,0 +1,370 @@
...
+
+def parse_plugin_toml_from_string(unparsed_plugin_toml_str: str, /, local: bool = False) -> List[Plugin]:
+    """Parses a plugin toml, given the string loaded in."""
+    doc = atoml.parse(unparsed_plugin_toml_str)
+    found_plugins: List[Plugin] = []
+    for plug_entry in doc["plugins"]:
+        if local:
+            enabled = plug_entry.get("enabled", True)
+        else:
+            enabled = None
```
[the remark this answers is not in the document]
Comment (onerandomusername) [reply modmail/addons/plugins.py:116]: Tbh I think this in particular looks worse when compressed.

File: modmail/bot.py [hunk 8]
```diff
@@ -171,31 +181,48 @@ async def close(self) -> None:
...
+            # no local plugins
+            pass
+        else:
+            for plug in self.installed_plugins:
+                if plug.enabled:
+                    continue
+                self.logger.debug(f"Not loading {plug.__str__()} on start since it's not enabled.")
+                dont_load_at_start.extend(plug.modules)
+
+        for plug in PLUGINS:
```
[the remark this answers is not in the document]
Comment (onerandomusername) [reply modmail/bot.py:243]: This is intentionally like this to be easier to connect to the database in the future.

TL;DR: all plugins need to have their installation reference stored in the base and that will be incorporated into this eventually.

File: modmail/constraints.txt [hunk 9]
```diff
@@ -0,0 +1,29 @@
+# NOTICE: This file is automatically generated by scripts/export_requirements.py
```
Comment (onerandomusername) [inline modmail/constraints.txt:1]: This file may not be needed. Currently discussing internally.

File: modmail/addons/converters.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (onerandomusername) [reply modmail/addons/converters.py]: resolved in https://github.com/discord-modmail/modmail/commit/9987b2665d86b0f3d550bd3837fb412c6f661f39

File: modmail/addons/converters.py [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+if TYPE_CHECKING:
+    from discord.ext.commands import Context
+
+    from modmail.log import ModmailLogger
+
+LOCAL_REGEX: re.Pattern = re.compile(r"^\@local (?P<addon>[^@\s]+)$")
+ZIP_REGEX: re.Pattern = re.compile(
+    r"^(?:https?:\/\/)?(?P<url>(?P<domain>.*\..+?)\/(?P<path>.*\.zip)) (?P<addon>[^@\s]+)$"
+)
+REPO_REGEX: re.Pattern = re.compile(
```
[the remark this answers is not in the document]
Comment (onerandomusername) [reply modmail/addons/converters.py]: resolved in https://github.com/discord-modmail/modmail/commit/9987b2665d86b0f3d550bd3837fb412c6f661f39

File: modmail/addons/errors.py [hunk 11]
```diff
@@ -0,0 +1,26 @@
...
+
+
+class PluginNotFoundError(PluginError):
+    """Plugins are not found and can therefore not be actioned on."""
+
+    pass
+
+
+class NoPluginTomlFoundError(PluginError):
+    """Raised when a plugin.toml file is expected to exist but does not exist."""
```
[the remark this answers is not in the document]
Comment (onerandomusername) [reply modmail/addons/errors.py:26]: resolved in https://github.com/discord-modmail/modmail/commit/9987b2665d86b0f3d550bd3837fb412c6f661f39

File: modmail/addons/helpers.py [hunk 12]
```diff
@@ -1,14 +1,24 @@
...
+
 from modmail.bot import ModmailBot
 from modmail.log import ModmailLogger
-from modmail.utils.cogs import BotModes, ExtMetadata, ModmailCog
+from modmail.utils.cogs import BOT_MODE, BotModeEnum, ExtMetadata
+from modmail.utils.cogs import ModmailCog as _ModmailCog
 
 
-__all__ = ["PluginCog", ModmailBot, ModmailLogger, BotModes, ExtMetadata]
+__all__ = [
```
[the remark this answers is not in the document]
Comment (onerandomusername) [reply modmail/addons/helpers.py]: resolved in https://github.com/discord-modmail/modmail/commit/9987b2665d86b0f3d550bd3837fb412c6f661f39

File: modmail/addons/models.py [hunk 13]
```diff
@@ -0,0 +1,193 @@
...
+        else:
+            self.name = name
+        self.folder_path = folder_path
+        self.min_bot_version = min_bot_version
+        self.local = local
+        self.enabled = enabled
+
+        self.dependencies = dependencies or []
+
+        self.modules = dict()
```
[the remark this answers is not in the document]
Comment (onerandomusername) [reply modmail/addons/models.py]: resolved in https://github.com/discord-modmail/modmail/commit/9987b2665d86b0f3d550bd3837fb412c6f661f39

File: modmail/addons/plugins.py [hunk 14]
```diff
@@ -0,0 +1,370 @@
+# original source:
+# https://github.com/python-discord/bot/blob/a8869b4d60512b173871c886321b261cbc4acca9/bot/utils/extensions.py
+# MIT License 2021 Python Discord
+"""
+Helper utililites for managing plugins.
```
[the remark this answers is not in the document]
Comment (onerandomusername) [reply modmail/addons/plugins.py]: resolved in https://github.com/discord-modmail/modmail/commit/9987b2665d86b0f3d550bd3837fb412c6f661f39

File: modmail/addons/models.py [hunk 15]
```diff
@@ -0,0 +1,193 @@
...
+        min_bot_version: Optional[str] = None,
+        name: Optional[str] = None,
+        **kw,
+    ):
+        self.folder_name = folder
+        self.description = description
+        if name is None:
+            self.name = self.folder_name
+        else:
+            self.name = name
```
[the remark this answers is not in the document]
Comment (onerandomusername) [reply modmail/addons/models.py]: resolved in https://github.com/discord-modmail/modmail/commit/9987b2665d86b0f3d550bd3837fb412c6f661f39

File: modmail/addons/plugins.py [hunk 16]
```diff
@@ -0,0 +1,370 @@
...
+    toml_plugin_names = [p.folder_name for p in toml_plugins]
+    for path in detection_path.iterdir():
+        logger.debug(f"detection_path / path: {path}")
+        if path.is_dir():
+            # use an existing toml plugin object
+            if path.name in toml_plugin_names:
+                for p in toml_plugins:
+                    if p.folder_name == path.name:
+                        p.folder_path = path
+                        all_plugins.add(p)
```
[the remark this answers is not in the document]
Comment (onerandomusername) [reply modmail/addons/plugins.py]: resolved in https://github.com/discord-modmail/modmail/commit/9987b2665d86b0f3d550bd3837fb412c6f661f39

File: modmail/addons/plugins.py [hunk 17]
```diff
@@ -0,0 +1,370 @@
...
+    if len(temp_direct_children) == 1:
+        folder = temp_direct_children[0]
+        if folder.is_dir():
+            addon_repo_path = addon_repo_path / folder
+    del temp_direct_children
+    # figure out which directory plugins are in. Both Plugins and plugins are supported.
+    # default is plugins.
+    plugin_directory = None
+    direct_children = [p for p in addon_repo_path.iterdir()]
+    logger.debug(f"{direct_children = }")
```
[the remark this answers is not in the document]
Comment (onerandomusername) [reply modmail/addons/plugins.py]: resolved in https://github.com/discord-modmail/modmail/commit/9987b2665d86b0f3d550bd3837fb412c6f661f39

File: modmail/addons/plugins.py [hunk 18]
```diff
@@ -0,0 +1,370 @@
...
+
+from modmail import plugins
+from modmail.addons.errors import NoPluginDirectoryError, NoPluginTomlFoundError
+from modmail.addons.models import Plugin
+from modmail.log import ModmailLogger
+from modmail.utils.cogs import ExtMetadata
+from modmail.utils.extensions import ModuleName, unqualify
+
+
+__all__ = [
```
[the remark this answers is not in the document]
Comment (onerandomusername) [reply modmail/addons/plugins.py]: resolved in https://github.com/discord-modmail/modmail/commit/9987b2665d86b0f3d550bd3837fb412c6f661f39

File: modmail/addons/plugins.py [hunk 19]
```diff
@@ -0,0 +1,370 @@
...
+        stderr=subprocess.PIPE,
+    )
+    stdout, stderr = await proc.communicate()
+    logger.debug(f"{stdout.decode() = }")
+
+    if stderr:
+        stderr = stderr.replace(PIP_NO_ROOT_WARNING, b"").strip()
+        if len(stderr.decode()) > 0:
+            logger.error(f"Received stderr: '{stderr.decode()}'")
+            raise Exception("Something went wrong when installing.")
```
[the remark this answers is not in the document]
Comment (onerandomusername) [reply modmail/addons/plugins.py]: resolved in https://github.com/discord-modmail/modmail/commit/9987b2665d86b0f3d550bd3837fb412c6f661f39

File: modmail/addons/utils.py [hunk 20]
```diff
@@ -0,0 +1,63 @@
...
+from modmail.addons.models import SourceTypeEnum
+from modmail.errors import HTTPError
+
+
+if TYPE_CHECKING:
+    from aiohttp import ClientSession
+
+    from modmail.addons.models import AddonSource
+    from modmail.log import ModmailLogger
+logger: ModmailLogger = logging.getLogger(__name__)
```
[the remark this answers is not in the document]
Comment (onerandomusername) [reply modmail/addons/utils.py:19]: resolved in https://github.com/discord-modmail/modmail/commit/9987b2665d86b0f3d550bd3837fb412c6f661f39

File: modmail/addons/plugins.py [hunk 21]
```diff
@@ -0,0 +1,370 @@
...
+PYTHON_INTERPRETER: Optional[str] = sys.executable
+
+PIP_NO_ROOT_WARNING = (
+    "WARNING: Running pip as the 'root' user can result in broken permissions and "
+    "conflicting behaviour with the system package manager. "
+    "It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv"
+).encode()
+
+
+async def install_dependencies(plugin: Plugin) -> str:
```
[the remark this answers is not in the document]
Comment (onerandomusername) [reply modmail/addons/plugins.py]: resolved in https://github.com/discord-modmail/modmail/commit/9987b2665d86b0f3d550bd3837fb412c6f661f39

File: modmail/bot.py [hunk 22]
```diff
@@ -36,14 +38,17 @@ class ModmailBot(commands.Bot):
...
     def __init__(self, **kwargs):
         self.config = CONFIG
-        self.start_time: t.Optional[arrow.Arrow] = None  # arrow.utcnow()
-        self.http_session: t.Optional[ClientSession] = None
+        self.start_time: Optional[arrow.Arrow] = None  # arrow.utcnow()
+        self.http_session: Optional[ClientSession] = None
         self.dispatcher = Dispatcher()
 
+        # keys: plugins, list values: all plugin files
+        self.installed_plugins: Set[Plugin] = {}
```
[the remark this answers is not in the document]
Comment (onerandomusername) [reply modmail/bot.py]: resolved in https://github.com/discord-modmail/modmail/commit/9987b2665d86b0f3d550bd3837fb412c6f661f39

File: modmail/addons/models.py [hunk 23]
```diff
@@ -0,0 +1,264 @@
...
+        arg_mapping: Dict[str, Plugin] = dict()
+        for plug in loaded_plugs:
+            for name in plug.name, plug.folder_name:
+                arg_mapping[name] = plug
+
+        result = process.extract(
+            argument,
+            arg_mapping.keys(),
+            scorer=fuzz.ratio,
+            score_cutoff=69,
```
[the remark this answers is not in the document]
Comment (onerandomusername) [reply modmail/addons/models.py]: resolved in https://github.com/discord-modmail/modmail/commit/9987b2665d86b0f3d550bd3837fb412c6f661f39

File: modmail/__init__.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (onerandomusername) [reply modmail/__init__.py]: resolved in https://github.com/discord-modmail/modmail/commit/9987b2665d86b0f3d550bd3837fb412c6f661f39

File: modmail/extensions/extension_manager.py [hunk 24]
```diff
@@ -286,11 +330,12 @@ def manage(self, action: Action, ext: str) -> t.Tuple[str, t.Optional[str]]:
 
             error_msg = f"{e.__class__.__name__}: {e}"
             msg = f":x: Failed to {verb} {self.type} `{ext}`:\n```\n{error_msg}```"
-        else:
-            msg = f":thumbsup: {self.type.capitalize()} successfully {verb}ed: `{ext}`."
+
+        if msg is None:
+            msg = f":thumbsup: {self.type.capitalize()} successfully {verb.rstrip('e')}ed: `{ext}`."
```
[the remark this answers is not in the document]
Comment (onerandomusername) [reply modmail/extensions/extension_manager.py:336]: its possible for msg to be None when an exception is raised, as Action.INSTALL does not set msg.
